Step 1

ilbender opened the pull request:
I'm mostly unsure about the usage of LiveData and where/should I should actually fetch the data

ilbender committed:
Fixes build issues.

ilbender committed:
implemented the first part:
1. created the interface in FeedApiService
2. created DTOs in DataTransferObjects.kt
3. implemented the interface in FeedApiResponseGenerator.kt
4. fetched and updated UI in FeedViewModel.kt -- most unsure about this

ilbender committed:
moved some constants around

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/network/FeedApiResponseGenerator.kt:
```diff
@@ -0,0 +1,22 @@
+
+    private val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
+
+    val retrofit = Retrofit.Builder().baseUrl(BASE_URL)
```
Looks like this property can be private.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/network/FeedApiResponseGenerator.kt:
```diff
@@ -0,0 +1,22 @@
+        .addConverterFactory(MoshiConverterFactory.create(moshi))
+        .addCallAdapterFactory(RxJava2CallAdapterFactory.create()).build()
+
+    fun createFeedApiService() : FeedApiService {
```
`FeedApiResponseGenerator` is a singleton, this is good for performance.
But each call to `createFeedApiService()` creates a new service.
How can we reuse the same service?

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt:
```diff
@@ -2,7 +2,20 @@ package com.lightricks.feedexercise.ui
+import androidx.lifecycle.ViewModelProvider
 import com.lightricks.feedexercise.R
+import com.lightricks.feedexercise.data.FeedItem
+import com.lightricks.feedexercise.network.FeedApiResponseGenerator
```
Unused imports?

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt:
```diff
@@ -3,3 +3,23 @@ package com.lightricks.feedexercise.network
+    val templatesMetadata : List<TemplatesMetadataItem>
+)
+
+object Constant{
```
```suggestion
object Constant {
```

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt:
```diff
@@ -11,5 +24,10 @@ class MainActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.main_activity)
+        supportFragmentManager.beginTransaction()
+            .replace(R.id.fragment_container_view,FeedFragment()).commit()
```
```suggestion
            .replace(R.id.fragment_container_view, FeedFragment()).commit()
```

bmax-ltx replied:
I don't think you need to do this. 
If you look at the layout (`main_activity.xml`), it should refer to `FeedFragment`.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,20 +1,60 @@
 open class FeedViewModel : ViewModel() {
+
+    init {
+        this.fetchFeedItems()
```
There is an init block below (line 71) that calls `refresh()`.
Let's either use it or get rid of it.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,20 +1,60 @@
     fun getIsLoading(): LiveData<Boolean> {
-        //todo: fix the implementation
-        return MutableLiveData()
+        return stateInternal.map { state -> state.isLoading }
```
👍

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,20 +1,60 @@
+        feedApiService.getFeedData()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ feedResponse ->
```
We should dispose the subscription in `onCleared()` (when the `ViewModel` is destroyed).

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,20 +1,60 @@
+            })
+    }
+
+    fun handleNetworkError(error: Throwable){
```
This is the place to post the `networkErrorEvent`.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,20 +1,60 @@
+    fun handleResponse(feedResponse: TemplatesMetadata) {
+        val rawResponseList = feedResponse.templatesMetadata
+        var feedItemLst : MutableList<FeedItem> = mutableListOf()
+        for(item in rawResponseList){
```
```suggestion
        for (item in rawResponseList){
```

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,20 +1,60 @@
+        val rawResponseList = feedResponse.templatesMetadata
+        var feedItemLst : MutableList<FeedItem> = mutableListOf()
+        for(item in rawResponseList){
+            feedItemLst.add(FeedItem(item.id, BASE_THUMBNAIL_URL + item.templateThumbnailURI,item.isPremium))
```
```suggestion
            feedItemLst.add(FeedItem(item.id, BASE_THUMBNAIL_URL + item.templateThumbnailURI, item.isPremium))
```

bmax-ltx replied:
It's good practice to extract the conversion logic to a function (or extension function).

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,20 +1,60 @@
+        for(item in rawResponseList){
+            feedItemLst.add(FeedItem(item.id, BASE_THUMBNAIL_URL + item.templateThumbnailURI,item.isPremium))
+        }
+        updateState{FeedViewModel.State(feedItemLst,false)}
```
```suggestion
        updateState { FeedViewModel.State(feedItemLst, false) }
```

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -23,8 +63,7 @@ open class FeedViewModel : ViewModel() {
     fun getFeedItems(): LiveData<List<FeedItem>> {
-        //todo: fix the implementation
-        return MutableLiveData()
+        return stateInternal.map { state -> state.feedItems ?: emptyList() }
```
Let's also implement `getIsEmpty()`. 
This is needed in order to show an empty view. See `feed_fragment.xml` (look for `empty_text`).

bmax-ltx reviewed (changes requested):
Great start!
I have some comments.

ilbender committed:
Update app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt

Co-authored-by: Baruch Maximenko <bmaximenko@lightricks.com>

ilbender committed:
Update app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

Co-authored-by: Baruch Maximenko <bmaximenko@lightricks.com>

ilbender committed:
fixed stuff according to PR including:
1. apiGenerator singleton
2. fixed disposable leak
3. extracted string concatenation logic to a function

ilbender committed:
forgot to add to last commit

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/network/FeedApiResponseGenerator.kt:
```diff
@@ -0,0 +1,25 @@
+    init {
+        feedApiService = retrofit.create(FeedApiService::class.java)
+    }
+    fun createFeedApiService() : FeedApiService {
```
Now the name `createFeedApiService` is misleading, because it doesn't always create a new service.
It would be better to define `feedApiService` as a `val`:
`val feedApiService = retrofit.create(FeedApiService::class.java)`
It will be initialized once and will return the same instance.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,30 +1,73 @@
+            }, { error ->
+                handleNetworkError(error)
+            })
+        compositeDisposable.add(subscribe)
```
👍

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,30 +1,73 @@
 
+    fun handleNetworkError(error: Throwable){
+        val msg : String = error.message ?: "Unexpected error"
+        networkErrorEvent.postValue(Event(msg))
```
👍

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -61,12 +104,16 @@ open class FeedViewModel : ViewModel() {
  */
-class FeedViewModelFactory : ViewModelProvider.Factory {
-    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+class FeedViewModelFactory (): ViewModelProvider.Factory {
```
I'm not sure you need the parentheses.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,30 +1,73 @@
+        val rawResponseList = feedResponse.templatesMetadata
+        val feedItemLst : MutableList<FeedItem> = mutableListOf()
+        for(item in rawResponseList){
+            val currItem = FeedItem(item.id, responseUrlToThumbnailUrl(item) ,item.isPremium)
```
It can be useful to make a function that converts `TemplatesMetadataItem` to `FeedItem` or 
`List< TemplatesMetadataItem >` to `List< FeedItem >` since you might use it in other places.

bmax-ltx reviewed (approved):
Looks even better!
A couple of minor suggestions.
